### PR TITLE
Uplift `clippy::{drop,forget}_{ref,copy}` lints

### PR DESCRIPTION
**Pull request by Urgau**
This PR aims at uplifting the `clippy::drop_ref`, `clippy::drop_copy`, `clippy::forget_ref` and `clippy::forget_copy` lints.

Those lints are/were declared in the correctness category of clippy because they lint on useless and most probably is not what the developer wanted.

## `drop_ref` and `forget_ref`

> :information_source: The lints have been renamed to `dropping_copy_types` and `forgetting_copy_types` in https://github.com/rust-lang/rust/pull/111756

The `drop_ref` and `forget_ref` lint checks for calls to `std::mem::drop` or `std::mem::forget` with a reference instead of an owned value.

### Example

```rust
let mut lock_guard = mutex.lock();
std::mem::drop(&lock_guard) // Should have been drop(lock_guard), mutex
// still locked
operation_that_requires_mutex_to_be_unlocked();
```

### Explanation

Calling `drop` or `forget` on a reference will only drop the reference itself, which is a no-op. It will not call the `drop` or `forget` method on the underlying referenced value, which is likely what was intended.

## `drop_copy` and `forget_copy`

> :information_source: The lints have been renamed to `dropping_references` and `forgetting_references` in https://github.com/rust-lang/rust/pull/111756

The `drop_copy` and `forget_copy` lint checks for calls to `std::mem::forget` or `std::mem::drop` with a value that derives the Copy trait.

### Example

```rust
let x: i32 = 42; // i32 implements Copy
std::mem::forget(x) // A copy of x is passed to the function, leaving the
                    // original unaffected
```

### Explanation

Calling `std::mem::forget` [does nothing for types that implement Copy](https://doc.rust-lang.org/std/mem/fn.drop.html) since the value will be copied and moved into the function on invocation.

-----

Followed the instructions for uplift a clippy describe here: https://github.com/rust-lang/rust/pull/99696#pullrequestreview-1134072751

cc @m-ou-se (as T-libs-api leader because the uplifting was discussed in a recent meeting)

### CODE REVIEW

File: compiler/rustc_middle/src/mir/visit.rs
Comment (Nilstrieb) [inline compiler/rustc_middle/src/mir/visit.rs:872]: could you keep the comment on the binding?

**Review by Noratrieb — COMMENTED** [review]
some cases in rustc, fun (can't wait for clippy to finally work on rustc^^)

File: src/tools/clippy/clippy_lints/src/drop_forget_ref.rs
Comment (Alexendoo) [inline src/tools/clippy/clippy_lints/src/drop_forget_ref.rs:212]: We'd want to keep these around but as an early return to avoid the extra cases appearing in e.g. `clippy::drop_non_drop`

https://github.com/rust-lang/rust/pull/109732/files#diff-ac6c6b2cb3727ed4c68d901f79187bfc83a4958814e261c425714e8319c29915R1-R13

File: compiler/rustc_infer/src/infer/nll_relate/mod.rs
Comment (Urgau) [reply compiler/rustc_infer/src/infer/nll_relate/mod.rs:912]: Probably yes, changed!

**Comment by Dragon-Hatcher** [comment]
After further inspection the terminal links standard is a little more involved. Let me make some fixes.

https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda

@rustbot label +S-waiting-on-author -S-waiting-on-review

**Comment by Dragon-Hatcher** [comment]
Oops. Wrong issue. Sorry!

@rustbot label -S-waiting-on-author +S-waiting-on-review

File: src/tools/clippy/tests/ui/deprecated.rs
Comment (Alexendoo) [inline src/tools/clippy/tests/ui/deprecated.rs:21]: Looks like this got left behind, it can be removed

Comment (Urgau) [reply src/tools/clippy/tests/ui/deprecated.rs]: Yeah, left over from the `cargo dev deprecate`. Should now be fixed.

**Comment by Urgau** [comment]
Nominating for T-lang approval, as it seems to be the norm.

@rustbot label: +I-lang-nominated

**Comment by nikomatsakis** [comment]
@rustbot labels -I-lang-nominated

@rfcbot fcp merge

We discussed this our @rust-lang/lang meeting and we agree that, when this occurs, it is very likely to be a bug or misunderstanding, which is our usual criteria for uplift.

**Comment by nikomatsakis** [comment]
@cramertj made the point that we should suggest people replace with `let _ = x` if the intent is to mark the value as used to disable the "unused variable" lint.

**Comment by nikomatsakis** [comment]
@rustbot labels -T-compiler -T-libs +T-lang

Whether or not to add a lint is a lang team decision, typically.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

We discussed this our @rust-lang/lang meeting and we agree that, when this occurs, it is very likely to be a bug or misunderstanding, which is our usual criteria for uplift. So proposing to FCP merge!

(Come on @rfcbot , you know you want to)

**Comment by camsteffen** [comment]
Should the "ref" and "copy" lints be collapsed? Refs are Copy.

**Comment by tialaramex** [comment]
> Should the "ref" and "copy" lints be collapsed? Refs are Copy.

While technically the Copy lints would trap this mistake, the goal is to tell the programmer when they did something that's almost certainly wrong, which will often mean the programmer misunderstood what's happening and the diagnostics should help them get a clearer picture in which they learn why this is wrong. The best diagnostics cause a reader to think "Oh! I'm so silly, I understand what I did wrong, I'll fix it" and the worst just cause confusion, "Huh? But why is that wrong? Rust is a stupid language"

If I drop(&goose) I probably *should* have written drop(goose) and so the diagnostic ought to focus on explaining that dropping references, specifically, doesn't drop the thing referred to, only the reference itself, which is useless. Maybe I should drop(goose).

On the other hand if I drop(goose) and goose is Copy, then the diagnostic wants to focus on explaining that because the goose is Copy this doesn't actually achieve anything - maybe I don't need to write anything at all here?

While rustc is capable of issuing more nuanced diagnostics for different situations in a single named error, programmers cannot choose to forbid/ allow/ etc. anything less coarse than a whole named error AFAIK, it seems plausible that a team might decide to forbid drop_ref for example and yet not forbid drop_copy.

**Comment by Urgau** [comment]
The PR has been rebased, new tests have been adjusted and the lint output has been improved (less space and `let _ = ...` suggestion).

@rustbot ready

*(and re-rolling since [TaKO8Ki](https://github.com/TaKO8Ki) doesn't seem active)*
r? compiler

**Review by davidtwco — APPROVED** [review]
LGTM, r=me after fixing one nit in a comment

**Comment by GuillaumeGomez** [comment]
r=davidtwco

**Comment by matthiaskrgr** [comment]
@bors r- rollup=iffy
https://github.com/rust-lang/rust/pull/110463#issuecomment-1512129233

**Comment by Urgau** [comment]
Rebased but still waiting on the stdarch update to be merged https://github.com/rust-lang/rust/pull/110285.

@rustbot labels: -S-waiting-on-author +S-blocked

**Comment by Urgau** [comment]
The stdarch update has been merged https://github.com/rust-lang/rust/pull/110285 and I've rebased the PR to fix the conflicts. The PR is ready again.

@rustbot labels: +S-waiting-on-review -S-blocked

**Comment by Urgau** [comment]
@jethrogb, in this code `entry` is of type `&SpinMutex<WaitEntry>` and through dropping him doesn't do anything. Can I just remove him ? or is there something else I should do? 
https://github.com/rust-lang/rust/blob/25444e5a2e7293dbb4932df1301b9be4222244fe/library/std/src/sys/sgx/waitqueue/mod.rs#L202-L216

**Comment by jethrogb** [comment]
@raoulstrackx do you think it should have been `drop(entry_guard)` instead maybe?

**Comment by jethrogb** [comment]
That at least seems to be in line with how the scoping works in notify_all

**Comment by jethrogb** [comment]
Although, in that last case, I guess it doesn't really work as intended, because it's dropping a copy of `entry`, not `entry` itself.

**Comment by jethrogb** [comment]
The [nomicon](https://doc.rust-lang.org/nomicon/unbounded-lifetimes.html) doesn't really have any good tips on how to otherwise bound an unbounded lifetime.

**Comment by raoulstrackx** [comment]
I think most likely we meant `drop(entry_guard)` here.

**Comment by Urgau** [comment]
Okay, I've changed it to `drop(entry_guard)`, let's see if it pass CI (I've temporarily enable the dist-various runner).

@rustbot labels: -A-testsuite -T-infra

**Comment by Urgau** [comment]
@jethrogb @raoulstrackx Any objection to landing this with `drop(entry_guard)` instead of `drop(entry)` ?

**Comment by jethrogb** [comment]
@raoulstrackx what do you think of this change (compared to master)

```diff
diff --git a/library/std/src/sys/sgx/waitqueue/mod.rs b/library/std/src/sys/sgx/waitqueue/mod.rs
index 61bb11d9a6f..3e3caa0d00f 100644
--- a/library/std/src/sys/sgx/waitqueue/mod.rs
+++ b/library/std/src/sys/sgx/waitqueue/mod.rs
@@ -202,12 +202,18 @@ pub fn wait_timeout<T, F: FnOnce()>(
     pub fn notify_one<T>(
         mut guard: SpinMutexGuard<'_, WaitVariable<T>>,
     ) -> Result<WaitGuard<'_, T>, SpinMutexGuard<'_, WaitVariable<T>>> {
+        // SAFETY: lifetime of the pop() return value is limited to map
+        // closure (The closure return value is 'static). The underlying
+        // stack frame won't be freed until after the WaitGuard created below
+        // is dropped.
         unsafe {
-            if let Some(entry) = guard.queue.inner.pop() {
+            let tcs = guard.queue.inner.pop().map(|entry| -> Tcs {
                 let mut entry_guard = entry.lock();
-                let tcs = entry_guard.tcs;
                 entry_guard.wake = true;
-                drop(entry);
+                entry_guard.tcs
+            });
+
+            if let Some(tcs) = tcs {
                 Ok(WaitGuard { mutex_guard: Some(guard), notified_tcs: NotifiedTcs::Single(tcs) })
             } else {
                 Err(guard)
@@ -223,6 +229,9 @@ pub fn notify_one<T>(
     pub fn notify_all<T>(
         mut guard: SpinMutexGuard<'_, WaitVariable<T>>,
     ) -> Result<WaitGuard<'_, T>, SpinMutexGuard<'_, WaitVariable<T>>> {
+        // SAFETY: lifetime of the pop() return values are limited to the
+        // while loop body. The underlying stack frames won't be freed until
+        // after the WaitGuard created below is dropped.
         unsafe {
             let mut count = 0;
             while let Some(entry) = guard.queue.inner.pop() {
@@ -230,6 +239,7 @@ pub fn notify_all<T>(
                 let mut entry_guard = entry.lock();
                 entry_guard.wake = true;
             }
+
             if let Some(count) = NonZeroUsize::new(count) {
                 Ok(WaitGuard { mutex_guard: Some(guard), notified_tcs: NotifiedTcs::All { count } })
             } else {
```

**Comment by raoulstrackx** [comment]
Yes I think that's an improvement!

**Comment by Urgau** [comment]
Alright, I've fixed two other compilation issues: a Redox one (just had to replace the `drop` by `let _ =`) and a SGX one (see the discussion above for the changes).

@rustbot ready

**Pull request merged**
